thread/common/omrthread.c

```diff
@@ -2121,14 +2145,30 @@ threadInternalExit(void)
 		}
 	}
 
-	GLOBAL_UNLOCK_SIMPLE(lib);
-
-	/* We couldn't clear TLS earlier because the debug version of
-	 * GLOBAL_UNLOCK_SIMPLE() uses omrthread_self().
+#if defined(THREAD_ASSERTS)
+	/* We can't clear TLS before unlocking the global monitor because
+	 * the debug version of GLOBAL_UNLOCK_SIMPLE() uses omrthread_self().
+	 *
+	 * This opens up a small window of time where the thread library
+	 * could be shut down before this thread has completed exitting.
+	 * Fetch the TLS key under lock to minimize the chance of the
+	 * key being destroyed by omrthread_shutdown.
+	 *
+	 * This is only an issue if THREAD_ASSERTS is enabled.
 	 */
+	{
+		TLSKEY tlsKey = lib->self_ptr;
+		GLOBAL_UNLOCK_SIMPLE(lib);
+		if (detached) {
+			TLS_SET(tlsKey, NULL);
+		}
+	}
+#else /* THREAD_ASSERTS */
 	if (detached) {
 		TLS_SET(lib->self_ptr, NULL);
```

Comment:
i think `TLS_SET/pthread_setspecific/TlsSetValue` are thread-safe. so, the TLS can be always cleared after releasing the global lock. the `ifdefs THREAD_ASSERTS` can be removed.

Reply:
This has nothing to do with thread safety - the comment clearly states why TLS can't be cleared in the assert case.

Reply:
i meant ... can't we just use the below code block for all scenarios. this way ... we can avoid duplicate code in the ifdefs. the new scope can be avoided if the TLS key is retrieved at the start of the function. 
```
	{
		TLSKEY tlsKey = lib->self_ptr;
		GLOBAL_UNLOCK_SIMPLE(lib);
		if (detached) {
			TLS_SET(tlsKey, NULL);
		}
	}
```

Reply:
I see what you're saying, but this is still an issue. Using the tlsKey after the unlock still leaves a tiny window where the shutdown could free the key before the TLS_SET takes place. If the key were somehow reallocated, we'd be NULLing someone else's data.